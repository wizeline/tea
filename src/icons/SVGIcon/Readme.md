
**Note:** When the size property is not specified or an invalid size is specified the `<Icon />` component will use the default size. 

The default size is currently set to `xsmall`.

```jsx
<div>
  <h3>Size Cases</h3>
  <table>
    <tbody>
      <tr>
        <th>default</th>
        <th>xlarge</th>
        <th>large</th>
        <th>medium</th>
        <th>small</th>
        <th>xsmall</th>
        <th>404</th>
      </tr>
      <tr>
        <td><SVGIcon name="plus" /></td>
        <td><SVGIcon name="plus" size="xlarge" /></td>
        <td><SVGIcon name="plus" size="large" /></td>
        <td><SVGIcon name="plus" size="medium" /></td>
        <td><SVGIcon name="plus" size="small" /></td>
        <td><SVGIcon name="plus" size="xsmall" /></td>
        <td><SVGIcon name="plus" size="MISSINGNO" /></td>
        <td>plus</td>
      </tr>
    </tbody>
  </table>
</div>
```

**Note:** When the name property is not specified or an invalid name is specified the `<Icon />` component will not render any component. 

In these cases `null` is returned.

```jsx
<div>
  <h3>Name Error Cases</h3>
  <table>
    <tbody>
      <tr>
        <th>undefined</th>
        <th>404</th>
      </tr>
      <tr>
        <td><SVGIcon /></td>
        <td><SVGIcon name="MISSINGNO" /></td>
      </tr>
    </tbody>
  </table>
</div>
```
