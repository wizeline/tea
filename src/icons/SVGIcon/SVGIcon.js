// @flow

import React from 'react';
import type { SVGIconProps } from './SVGIconTypes';
import constants from './SVGIconConstants';
import SVGIconDivStyled from './styled/div';
import SVGIconSvgStyled from './styled/svg';
import getSVG from './data';
import defaultIconTheme from '../../themes/defaultTheme/components/Icon';

const SVGIcon = (props: SVGIconProps) => {
  const {
    name,
    top,
    color,
    svgData,
    invert,
    xmall,
    small,
    medium,
    large,
    xlarge,
  } = props;
  const fillColor = invert && !color ? undefined : color;
  const svg = svgData || getSVG(name);
  if (!svg) return null;
  const svgViewBox = svg.props.viewBox || constants.defaultViewBox;
  const svgChildren = svg.props.children;
  return (
    <SVGIconDivStyled
      top={top}
      xmall={xmall}
      small={small}
      medium={medium}
      large={large}
      xlarge={xlarge}
    >
      <SVGIconSvgStyled
        invert={invert}
        fill={fillColor}
        viewBox={svgViewBox}
        preserveAspectRatio={constants.preserveAspectRatio}
      >
        {svgChildren}
      </SVGIconSvgStyled>
    </SVGIconDivStyled>
  );
};

SVGIcon.defaultProps = {
  theme: defaultIconTheme,
};

export default SVGIcon;
