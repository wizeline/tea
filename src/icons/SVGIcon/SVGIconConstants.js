// @flow

const constants: {
  defaultTop: number,
  preserveAspectRatio: string,
  defaultViewBox: string,
  defaultFillColorThemeName: string,
  defaultSizeKey: string,
  sizes: Object,
} = {
  defaultTop: 0,
  preserveAspectRatio: 'xMidYMid meet',
  defaultViewBox: '0 0 24 24',
  defaultFillColorThemeName: 'textPrimary',
  defaultSizeKey: 'xsmall',
  sizes: {
    xlarge: 56,
    large: 48,
    medium: 32,
    small: 24,
    xsmall: 16,
  },
};

export default constants;
