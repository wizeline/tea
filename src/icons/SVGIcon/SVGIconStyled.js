// @flow

/* eslint flowtype-errors/enforce-min-coverage: 0 */

import styled from 'styled-components';
import type {
  SVGFillProps,
  ContainerSizeProps,
  ContainerTopProps,
} from './SVGIconTypes';
import constants from './SVGIconConstants';

const getSVGFill = ({
  iconFillColor: iconFillColor = constants.defaultFillColor,
}: SVGFillProps) => iconFillColor;

const getContainerTop = ({ top }: ContainerTopProps) => top;
const getContainerSize = ({ size }: ContainerSizeProps) =>
  constants.sizes[size] || constants.sizes[constants.defaultSizeKey];

const SVGIconDivStyled = styled.div`
  display: inline-block;
  position: relative;
  height: ${getContainerSize}px;
  width: ${getContainerSize}px;
  top: ${getContainerTop}px;
`;

const SVGIconSvgStyled = styled.svg`
  fill: ${getSVGFill};
  user-select: none;
  margin: 0 auto;
`;

export { SVGIconDivStyled, SVGIconSvgStyled };
