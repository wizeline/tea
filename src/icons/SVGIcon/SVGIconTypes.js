// @flow

export type SVGFillProps = {
  iconFillColor: string,
  theme: {
    textPrimaryInvert: string,
  },
};

export type SVGIconProps = {
  name?: string,
  size?: string,
  color?: string,
  top?: number,
  svgData?: React.Node,
};

export type ContainerSizeProps = {
  size: number,
};

export type ContainerTopProps = {
  top: number,
};
