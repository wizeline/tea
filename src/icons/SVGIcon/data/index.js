// @flow

import addItem from './add-item';
import addPeople from './add-people';
import arrowDown from './arrow-down';
import bell from './bell';
import book from './book';
import building from './building';
import calendar from './calendar';
import check from './check';
import creditCard from './credit-card';
import cross from './cross';
import dotsH from './dots-h';
import duplicate from './duplicate';
import flag from './flag';
import heart from './heart';
import info from './info';
import lock from './lock';
import magnifier from './magnifier';
import pencil from './pencil';
import people from './people';
import plus from './plus';
import trash from './trash';

const lispToPascalCase = (word: string) =>
  word
    .split('-')
    .reduce(
      (result, token) =>
        result + token.replace(/^\w/, chr => chr.toUpperCase()),
    );

const data = {
  addItem,
  addPeople,
  arrowDown,
  bell,
  book,
  building,
  calendar,
  check,
  creditCard,
  cross,
  dotsH,
  duplicate,
  flag,
  heart,
  info,
  lock,
  magnifier,
  pencil,
  people,
  plus,
  trash,
};

const getSVG = (name: string) =>
  name === undefined ? null : data[lispToPascalCase(name)];

export default getSVG;
