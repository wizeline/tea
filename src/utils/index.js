// @flow

import supportSizes from './sizes/supportSizes';
import {
  arrayToObject,
  cssProperties,
  lispToCamelCase,
  propNames,
  propOrTheme,
  supportBooleanNameGroup,
  supportOr,
  supportOrTheme,
  toPixels,
  warn,
} from './coreUtils';

export {
  arrayToObject,
  cssProperties,
  lispToCamelCase,
  propNames,
  propOrTheme,
  supportBooleanNameGroup,
  supportOr,
  supportOrTheme,
  supportSizes,
  toPixels,
  warn,
};
