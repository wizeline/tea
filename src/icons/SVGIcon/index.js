// @flow

import withTheme from '../../utils/withTheme';
import SVGIcon from './SVGIcon';

export default withTheme(SVGIcon);
