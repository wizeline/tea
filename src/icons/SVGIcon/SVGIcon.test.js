import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import SVGIcon from './SVGIcon';
import SVGIconDivStyled from './styled/div';
import SVGIconSvgStyled from './styled/svg';
import { shallowExpectInvalid } from '../../utils/testHelpers';

describe('SVGIcon', () => {
  it('renders correctly with an undefined size and a valid name that references an SVG with a defined viewBox', () => {
    const wrapper = shallow(<SVGIcon name="cross" />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('top', '0px');
  });
  it('renders correctly with a valid size and a valid name that references an SVG without a defined viewBox', () => {
    const wrapper = shallow(
      <SVGIcon name="missing-view-box" size="large" top={20} />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('top', '20px');
  });
  it('renders nothing with an invalid name and undefined size attribute', () => {
    shallowExpectInvalid(<SVGIcon name="InvalidIconName" />);
  });
  it('renders nothing with invalid name and invalid size', () => {
    shallowExpectInvalid(<SVGIcon name="InvalidIconName" size="InvalidSize" />);
  });
  it('renders nothing with an undefined name and undefined size', () => {
    shallowExpectInvalid(<SVGIcon />);
  });
});

describe('SVGIconSvgStyled', () => {
  it('renders correctly without any defined attributes', () => {
    const wrapper = shallow(<SVGIconSvgStyled />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('renders correctly with defined color', () => {
    const wrapper = shallow(<SVGIconSvgStyled fill="red" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('SVGIconDivStyled', () => {
  it('renders correctly with defined valid container size', () => {
    const wrapper = shallow(<SVGIconDivStyled size="large" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('renders correctly with defined invalid container size', () => {
    const wrapper = shallow(<SVGIconDivStyled size="invalidSize" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('renders correctly with defined top', () => {
    const wrapper = shallow(<SVGIconDivStyled top={10} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
