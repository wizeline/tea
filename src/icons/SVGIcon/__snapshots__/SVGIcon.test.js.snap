// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVGIcon renders correctly when given svgData 1`] = `
<styled.div>
  <styled.svg
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 24 24"
  />
</styled.div>
`;

exports[`SVGIcon renders correctly with a valid size and a valid name that references an SVG without a defined viewBox 1`] = `
<styled.div
  size="large"
  top={20}
>
  <styled.svg
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 24 24"
  >
    <g
      fillRule="evenodd"
      transform="translate(0 2)"
    >
      <path
        d="M14.4 0H1.6C.712 0 .008.712.008 1.6L0 11.2c0 .888.712 1.6 1.6 1.6h12.8c.888 0 1.6-.712 1.6-1.6V1.6c0-.888-.712-1.6-1.6-1.6zm0 11.2H1.6V6.4h12.8v4.8zm0-8H1.6V1.6h12.8v1.6z"
      />
    </g>
  </styled.svg>
</styled.div>
`;

exports[`SVGIcon renders correctly with an undefined size and a valid name that references an SVG with a defined viewBox 1`] = `
<styled.div>
  <styled.svg
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 16 16"
  >
    <g
      fillRule="evenodd"
      transform="translate(1 1)"
    >
      <path
        d="M7 5.77L12.77 0 14 1.23 8.23 7 14 12.77 12.77 14 7 8.23 1.23 14 0 12.77 5.77 7 0 1.23 1.23 0 7 5.77z"
      />
    </g>
  </styled.svg>
</styled.div>
`;

exports[`SVGIcon renders nothing with an invalid name and undefined size attribute 1`] = `""`;

exports[`SVGIcon renders nothing with an undefined name and undefined size 1`] = `""`;

exports[`SVGIcon renders nothing with invalid name and invalid size 1`] = `""`;

exports[`SVGIconDivStyled renders correctly with defined invalid container size 1`] = `
.c0 {
  display: inline-block;
  position: relative;
  width: 16px;
  height: 16px;
  top: 0px;
}

<div
  className="c0"
  size="invalidSize"
/>
`;

exports[`SVGIconDivStyled renders correctly with defined top 1`] = `
.c0 {
  display: inline-block;
  position: relative;
  width: 16px;
  height: 16px;
  top: 10px;
}

<div
  className="c0"
/>
`;

exports[`SVGIconDivStyled renders correctly with defined valid container size 1`] = `
.c0 {
  display: inline-block;
  position: relative;
  width: 48px;
  height: 48px;
  top: 0px;
}

<div
  className="c0"
  size="large"
/>
`;

exports[`SVGIconSvgStyled renders correctly with defined color 1`] = `
.c0 {
  fill: red;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin: 0 auto;
}

<svg
  className="c0"
  fill="red"
/>
`;

exports[`SVGIconSvgStyled renders correctly without any defined attributes 1`] = `
.c0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin: 0 auto;
}

<svg
  className="c0"
/>
`;
